File: modules/expo-dual-camera/android/src/main/java/expo/modules/dualcamera/ExpoDualCameraModule.kt
package expo.modules.dualcamera

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoDualCameraModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoDualCamera")

    // Android: dual capture is not supported by this module (see ExpoDualCameraView).
    Constants(
      "isSupported" to false
    )

    View(ExpoDualCameraView::class) {
      Events("onInitError")
      Prop("mirrorFront") { _: ExpoDualCameraView, _: Boolean -> }
      Prop("layout") { _: ExpoDualCameraView, _: String -> }
    }
  }
}

File: modules/expo-dual-camera/ios/ExpoDualCameraModule.swift
import ExpoModulesCore
import AVFoundation

public class ExpoDualCameraModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoDualCamera")

    // Surfaced to JS as ExpoDualCamera.isSupported — gates the "Dual" camera mode.
    Constants([
      "isSupported": AVCaptureMultiCamSession.isMultiCamSupported
    ])

    View(ExpoDualCameraView.self) {
      Events("onInitError")

      Prop("mirrorFront") { (view: ExpoDualCameraView, value: Bool) in
        view.setMirrorFront(value)
      }

      // Lens arrangement: "sideBySide" (back|front halves) or "pip" (front bubble over back).
      Prop("layout") { (view: ExpoDualCameraView, value: String) in
        view.setLayout(value)
      }

      AsyncFunction("capturePhoto") { (view: ExpoDualCameraView, promise: Promise) in
        view.capturePhoto(promise)
      }
    }
  }
}
